File: app/src/main/java/com/example/medicineapplication/adapter/MedicineAdapter.java
package com.example.medicineapplication.adapter;

import android.content.Context;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.cardview.widget.CardView;
import androidx.recyclerview.widget.RecyclerView;

import com.example.medicineapplication.R;
import com.example.medicineapplication.model.Medicine;
import com.example.medicineapplication.model.Pharmacy;

import java.util.List;

public class MedicineAdapter extends RecyclerView.Adapter<MedicineAdapter.ViewHolder> {
    private final List<Medicine> mData;
    private final LayoutInflater inflater;
    private final MedicineAdapter.ItemClickListener itemClickListener;
    Context context;

    public MedicineAdapter(Context context, List<Medicine> data, MedicineAdapter.ItemClickListener onClick) {
        this.inflater = LayoutInflater.from(context);
        this.mData = data;
        this.context = context;
        this.itemClickListener = onClick;
    }

    @NonNull
    @Override
    public MedicineAdapter.ViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        View view = inflater.inflate(R.layout.medicine_item, parent, false);
        return new MedicineAdapter.ViewHolder(view);
    }

    @Override
    public void onBindViewHolder(@NonNull MedicineAdapter.ViewHolder holder, int position) {
        var medicineData = mData.get(position);
        Log.e("TAG", medicineData.getMedicineName());
        Log.e("TAG", medicineData.getDescription());
        Log.e("TAG", medicineData.getId());
        Log.e("TAG", medicineData.getPrice());
        Log.e("TAG", medicineData.getMedicineImage() + "");
        holder.txtMedicineName.setText(medicineData.getMedicineName());
        holder.txtMedicineName.setText(medicineData.getMedicineName());
        holder.medicineImg.setImageResource(medicineData.getMedicineImage());
        holder.txtPrice.setText(medicineData.getPrice());
        holder.txtMedicineDescription.setText(medicineData.getDescription());
        holder.container.setOnClickListener(v -> {
                    try {
                        itemClickListener.onItemClickMedicine(holder.getAdapterPosition(),
                                medicineData.getId());
                    } catch (Exception e) {
                        e.printStackTrace();
                    }
                }
        );
    }

    @Override
    public int getItemCount() {
        return mData.size();
    }

    public static class ViewHolder extends RecyclerView.ViewHolder implements View.OnClickListener {
        TextView txtMedicineName;
        ImageView medicineImg;
        TextView txtMedicineDescription;

        TextView txtPrice;
        CardView container;

        ViewHolder(View itemView) {
            super(itemView);
            this.txtMedicineName = itemView.findViewById(R.id.medicine_name);
            this.txtMedicineDescription = itemView.findViewById(R.id.medicine_description);
            this.txtPrice = itemView.findViewById(R.id.price);
            this.medicineImg = itemView.findViewById(R.id.medicine_img);
            this.container = itemView.findViewById(R.id.medicine_card);
            itemView.setOnClickListener(this);
        }

        @Override
        public void onClick(View v) {

        }

    }

    public interface ItemClickListener {
        void onItemClickMedicine(int position, String id);
    }

}

File: app/src/main/java/com/example/medicineapplication/adapter/MedicineAdapter.kt
package com.example.medicineapplication.adapter

import android.app.Activity
import android.view.ViewGroup
import androidx.recyclerview.widget.RecyclerView
import com.bumptech.glide.Glide
import com.example.medicineapplication.R
import com.example.medicineapplication.databinding.CategoryMedicineItemBinding
import com.example.medicineapplication.databinding.MedicineItemBinding
import com.example.medicineapplication.model.Pharmacy
import com.example.medicineapplication.model.Treatment

class MedicineAdapter(
    private var activity: Activity,
    var data: ArrayList<Treatment>,
    private var itemClickListener: ItemClickListener,

    ) : RecyclerView.Adapter<RecyclerView.ViewHolder>() {
    private val backgroundColor = listOf(
        R.color.mid_green,
        R.color.light_beige,
        R.color.dark_beige,
        R.color.light_orange,
        R.color.purple2,
        R.color.dark_yellow_green
    )

    companion object {
        private const val TYPE_HOME = 0
        private const val TYPE_CATEGORY = 1
    }

    class ViewHolderHome(var binding: MedicineItemBinding) : RecyclerView.ViewHolder(binding.root)
    class ViewHolderCategory(var binding: CategoryMedicineItemBinding) :
        RecyclerView.ViewHolder(binding.root)

    interface ItemClickListener {
        fun onItemClickMedicine(position: Int, id: String)
        fun onAddMedicineToFavorite(medicineId: Int)
    }

    override fun getItemViewType(position: Int): Int {
        return if (data[position].isFeatured == true) TYPE_CATEGORY else TYPE_HOME
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): RecyclerView.ViewHolder {
        return if (viewType == TYPE_HOME) {
            val binding = MedicineItemBinding.inflate(activity.layoutInflater, parent, false)
            ViewHolderHome(binding)
        } else {
            val binding =
                CategoryMedicineItemBinding.inflate(activity.layoutInflater, parent, false)
            ViewHolderCategory(binding)
        }
    }

    override fun onBindViewHolder(holder: RecyclerView.ViewHolder, position: Int) {
        val item = data[position]
        when (holder) {
            is ViewHolderHome -> {
                holder.binding.medicineName.text = item.name
                holder.binding.medicineDescription.text = item.description

                // تحميل صورة الدواء
                Glide.with(activity)
                    .load(item.image)
                    .placeholder(R.drawable.medicine_img)
                    .into(holder.binding.medicineImg)

                // حالة المفضلة
                if (item.is_favorite == true) {
                    holder.binding.favoriteImg.setImageResource(R.drawable.red_favorite)
                } else {
                    holder.binding.favoriteImg.setImageResource(R.drawable.favorite)
                }



                holder.binding.favoriteImg.setOnClickListener {
                    itemClickListener.onAddMedicineToFavorite(item.id)
                }

                holder.binding.root.setOnClickListener {
                    try {
                        itemClickListener.onItemClickMedicine(position, item.id.toString())
                    } catch (e: Exception) {
                        e.printStackTrace()
                    }
                }
            }

            is ViewHolderCategory -> {
                val colorResId = backgroundColor[position % backgroundColor.size]
                holder.binding.txtMedicineName.text = item.name
                holder.binding.txtMedicineDescription.text = item.description
                holder.binding.medicineCard.setCardBackgroundColor(activity.getColor(colorResId))

                // تحميل الصورة
                Glide.with(activity)
                    .load(item.image)
                    .placeholder(R.drawable.medicine_img)
                    .into(holder.binding.categoryImage)

                // حالة المفضلة
                holder.binding.favoriteImg.setImageResource(
                    if (item.is_favorite == true) R.drawable.red_favorite else R.drawable.favorite
                )

                holder.binding.favoriteImg.setOnClickListener {
                    itemClickListener.onAddMedicineToFavorite(item.id)
                }

                holder.binding.root.setOnClickListener {
                    try {
                        itemClickListener.onItemClickMedicine(position, item.id.toString())
                    } catch (e: Exception) {
                        e.printStackTrace()
                    }
                }
            }
        }

    }

    fun updateData(newList: List<Treatment>) {
        data.clear()
        data.addAll(newList)
        notifyDataSetChanged()
    }

    override fun getItemCount(): Int {
        return data.size
    }
}